apis/components/v1/modelcontroller_types.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"github.com/opendatahub-io/opendatahub-operator/v2/apis/components"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// shared by kserve and modelmeshserving
+	// value should match whats set in the XValidation below
+	ModelControllerInstanceName = "default-modelcontroller"
+	ModelControllerKind         = "ModelController"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope=Cluster
+// +kubebuilder:validation:XValidation:rule="self.metadata.name == 'default-modelcontroller'",message="ModelController name must be default-modelcontroller"
+// +kubebuilder:printcolumn:name="Ready",type=string,JSONPath=`.status.conditions[?(@.type=="Ready")].status`,description="Ready"
+// +kubebuilder:printcolumn:name="Reason",type=string,JSONPath=`.status.conditions[?(@.type=="Ready")].reason`,description="Reason"
+
+// ModelController is the Schema for the modelcontroller API, it is a shared component between kserve and modelmeshserving
+type ModelController struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ModelControllerSpec   `json:"spec,omitempty"`
+	Status ModelControllerStatus `json:"status,omitempty"`
+}
+
+// ModelControllerSpec defines the desired state of ModelController
+type ModelControllerSpec struct {
+	components.DevFlagsSpec `json:",inline"`
+	ModelMeshServing        operatorv1.ManagementState `json:"modelMeshServing,omitempty"`
+	Kserve                  operatorv1.ManagementState `json:"kserve,omitempty"`
```

Comment:
I'd rather create some kserver and modelmesh struct where we can add a `managementState` field for consistentcy, something like:

```go
type ModelControllerKerveSpec struct{
    components.DevFlagsSpec `json:",inline"`
    ManagementState operatorv1.ManagementState `json:"modelMeshServing,omitempty"`
}

type ModelControllerModelMeshSpec struct{
    components.DevFlagsSpec `json:",inline"`
    ManagementState operatorv1.ManagementState `json:"modelMeshServing,omitempty"`
}

type ModelControllerSpec struct {
        ModelMeshServing        *ModelControllerModelMeshSpec`json:"modelMeshServing,omitempty"`
	Kserve                  *ModelControllerKerveSpec `json:"kserve,omitempty"`
}
```

Reply:
Since we don't have a user facing API for the model controller, dev flags should probably be inherited from the kserve or model mesh components